# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

In [1]:
import pandas as pd
import numpy as np

In [2]:
original_data=pd.read_csv('e_commerce.csv')
original_data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom


## 评估数据

### 评估数据结构整洁度

In [3]:
original_data.sample(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
292511,562553,82580,BATHROOM METAL SIGN,1,8/5/2011 16:34,1.25,NaN,United Kingdom
63393,541567,21578,WOODLAND DESIGN COTTON TOTE BAG,6,1/19/2011 11:51,2.25,12681.0,France
481315,577358,22154,ANGEL DECORATION 3 BUTTONS,1,11/18/2011 15:59,0.83,NaN,United Kingdom
281662,561585,21977,PACK OF 60 PINK PAISLEY CAKE CASES,24,7/28/2011 10:46,0.55,13728.0,United Kingdom
502401,578844,22666,RECIPE BOX PANTRY YELLOW DESIGN,3,11/25/2011 16:14,5.79,NaN,United Kingdom
414639,572475,22158,3 HEARTS HANGING DECORATION RUSTIC,2,10/24/2011 13:37,2.95,15271.0,United Kingdom
309027,564061,82551,LAUNDRY 15C METAL SIGN,2,8/22/2011 14:38,1.45,14505.0,United Kingdom
225229,556636,90086,CRYSTAL FROG PHONE CHARM,1,6/13/2011 15:30,1.63,NaN,United Kingdom
176410,552000,21090,SET/6 COLLAGE PAPER PLATES,39,5/5/2011 15:56,0.39,NaN,United Kingdom
485088,577550,20996,JAZZ HEARTS ADDRESS BOOK,24,11/20/2011 15:05,0.19,17738.0,United Kingdom


从随机的10行数据中可以判断，每一行为每张发票中每种产品代码的的交易情况，每一列为每个交易情况的相关变量，结构上符合每行只有一个观察值，每列只有一个变量，每个单元格为一个观察值的整洁数据的特点。

### 评估数据内容干净度

评估数据内容的干净程度，主要从空缺值、重复数据、不一致数据、无效/错误数据等方面评估。

In [4]:
original_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  str    
 1   StockCode    541909 non-null  str    
 2   Description  540455 non-null  str    
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  str    
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  str    
dtypes: float64(2), int64(1), str(5)
memory usage: 33.1 MB


通过info方法得知，共有541909条观察值，其中Description、CustomerID存在空缺数据，InvoiceDate的数据类型应为日期时间，CustomerID数据类型应为字符串，需要做数据类型转换。

#### 评估空缺数据

将Description列空缺的行提取出来。

In [5]:
original_data[original_data['Description'].isnull()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/2010 11:52,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,12/1/2010 14:32,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,12/1/2010 14:33,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,12/1/2010 14:34,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535322,581199,84581,NaN,-2,12/7/2011 18:26,0.0,NaN,United Kingdom
535326,581203,23406,NaN,15,12/7/2011 18:31,0.0,NaN,United Kingdom
535332,581209,21620,NaN,6,12/7/2011 18:35,0.0,NaN,United Kingdom
536981,581234,72817,NaN,27,12/8/2011 10:33,0.0,NaN,United Kingdom


共筛选出来1454行，观察发现，这些Description列为空值的观察值中，UnitPrice的值似乎都为0，为了验证猜测，进一步通过条件筛选，看看是否有Description值为空，且UnitPrice值不为0的行。

In [6]:
original_data[(original_data['Description'].isnull()) & (original_data['UnitPrice']!=0)]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


条件筛选结果为空，因此猜想成立，即Description列为空值的观察值，它们的UnitPrice的值都为0。

而此次分析的目的是挖掘畅销产品，而产品名称Description和产品单价UnitPrice都是对分析有重要意义的变量，因此当两者同时缺少或无效时，对后续分析无意义，因此需要将Description为空的行删除。

将CustomerID列空缺的行提取出来。

In [7]:
original_data[original_data['CustomerID'].isnull()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,12/1/2010 11:52,0.00,NaN,United Kingdom
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,12/1/2010 14:32,2.51,NaN,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,12/1/2010 14:32,2.51,NaN,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,12/1/2010 14:32,0.85,NaN,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,12/1/2010 14:32,1.66,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
541536,581498,85099B,JUMBO BAG RED RETROSPOT,5,12/9/2011 10:26,4.13,NaN,United Kingdom
541537,581498,85099C,JUMBO BAG BAROQUE BLACK WHITE,4,12/9/2011 10:26,4.13,NaN,United Kingdom
541538,581498,85150,LADIES & GENTLEMEN METAL SIGN,1,12/9/2011 10:26,4.96,NaN,United Kingdom
541539,581498,85174,S/4 CACTI CANDLES,1,12/9/2011 10:26,10.79,NaN,United Kingdom


共提取出135080行CustomerID为空值的观察值。这些行除了CustomerID外，其他列都为有效数据，且CustomerID指代客户编号，空缺后对实际销售数据分析无影响，因此无需对CustomerID为空的行做处理。

#### 评估重复数据

从各列的意义来看，虽然InvoiceNo、StockCode、CustomerID均为交易、产品、客户的唯一标识符，但同一交易允许存在不同产品，同一产品可存在不同交易中，同一客户可一次下单不同产品或多次下单，因此这三个变量均可存在重复。

#### 评估不一致数据

不一致数据是指用不同的值指代同一目标，数据中Country变量有可能存在不一致数据。

In [8]:
original_data['Country'].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


在Country列的所有值中，United Kingdom、UK、U.K.都指代英国，USA、United States都指代美国，因此需要将UK、U.K.统一替换成United Kingdom，将USA替换程United States。

#### 评估无效/错误数据

使用describe方法检查数值数据的整体描述。

In [9]:
original_data.describe()

,Quantity,UnitPrice,CustomerID
count,541909.000000,541909.000000,406829.000000
mean,9.552250,4.611114,15287.690570
std,218.081158,96.759853,1713.600303
min,-80995.000000,-11062.060000,12346.000000
25%,1.000000,1.250000,13953.000000
50%,3.000000,2.080000,15152.000000
75%,10.000000,4.130000,16791.000000
max,80995.000000,38970.000000,18287.000000


Quantity和UnitPrice的最小值均为负数，不符合实际情况，因此这两列含有无效或错误数据。

通过条件筛选出Quantity列无效/错误数据的行。

In [10]:
original_data[original_data['Quantity']<0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,12/1/2010 9:41,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,12/1/2010 9:49,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,12/1/2010 10:24,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,12/1/2010 10:24,0.29,17548.0,United Kingdom
...,...,...,...,...,...,...,...,...
540449,C581490,23144,ZINC T-LIGHT HOLDER STARS SMALL,-11,12/9/2011 9:57,0.83,14397.0,United Kingdom
541541,C581499,M,Manual,-1,12/9/2011 10:28,224.69,15498.0,United Kingdom
541715,C581568,21258,VICTORIAN SEWING BOX LARGE,-5,12/9/2011 11:57,10.95,15311.0,United Kingdom
541716,C581569,84978,HANGING HEART JAR T-LIGHT HOLDER,-1,12/9/2011 11:58,1.25,17315.0,United Kingdom


共有10624 rows行的Quantity变量值为负数。通过观察发现这些行似乎InvoiceNo都以“C”开头，表示取消订单，为验证猜想，需要通过条件筛选，检查是否有Quantity为负数，而InvoiceNo不以C开头的行。

In [11]:
original_data[(original_data['InvoiceNo'].str[0]!='C')&(original_data['Quantity']<0)]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
2406,536589,21777,NaN,-10,12/1/2010 16:50,0.0,NaN,United Kingdom
4347,536764,84952C,NaN,-38,12/2/2010 14:42,0.0,NaN,United Kingdom
7188,536996,22712,NaN,-20,12/3/2010 15:30,0.0,NaN,United Kingdom
7189,536997,22028,NaN,-20,12/3/2010 15:30,0.0,NaN,United Kingdom
7190,536998,85067,NaN,-6,12/3/2010 15:30,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535333,581210,23395,check,-26,12/7/2011 18:36,0.0,NaN,United Kingdom
535335,581212,22578,lost,-1050,12/7/2011 18:38,0.0,NaN,United Kingdom
535336,581213,22576,check,-30,12/7/2011 18:38,0.0,NaN,United Kingdom
536908,581226,23090,missing,-338,12/8/2011 9:56,0.0,NaN,United Kingdom


有1336行满足筛选条件，因此猜想不成立。但发现在这些行里面，UnitPrice的似乎值都为0，为验证猜想，需要进一步对其进行条件筛选，检查其中是否有UnitPrice不为0的行。

In [12]:
original_data[(original_data['InvoiceNo'].str[0]!='C')&(original_data['Quantity']<0)&(original_data['UnitPrice']!=0)]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country


筛选结果为空，因此猜想成立，即Quantity为负数且InvoiceNo不以C开头的观察值中UnitPrice值全部为0。

因此Quantity为负数的情况有如下两种：
- InvoiceNo以C开头，表示订单取消
- UnitPrice为0，即单价为0英镑  

这些数据均不是有效成交数据，不贡献销售，因此对分析畅销产品无意义，所以后续数据清洗环节需要将Quantity为负数的数据删除。

通过条件筛选出UnitPrice列无效/错误数据的行。

In [13]:
original_data[original_data['UnitPrice']<0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
299983,A563186,B,Adjust bad debt,1,8/12/2011 14:51,-11062.06,NaN,United Kingdom
299984,A563187,B,Adjust bad debt,1,8/12/2011 14:52,-11062.06,NaN,United Kingdom


共有2个UnitPrice值为负数的观察值，从Description的含义来看，为调整坏账，并非有效销售数据，对后续的销售情况分析无意义，因此需要在数据清洗环节删除。

## 清理数据

从前面的数据评估环节得知，需要将数据做如下清洗：
- 将Description列为空的行删除
- 将UK、U.K.统一替换成United Kingdom
- 将USA替换程United States
- 将Quantity为负数的数据删除
- 将UnitPrice为负数的数据删除
- 将InvoiceDate的数据类型转换为日期时间
- 将CustomerID的数据类型转换为字符串

在清洗之前，需要将原始数据复制给新的数据cleaned_data，这样后续所有的清洗步骤均操作在复制出来的副本cleaned_data上。

In [14]:
cleaned_data = original_data.copy()
cleaned_data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom


#### 删除Description列为空的行

删除Description列为空的行，并检查删除后该列空缺值的个数是否为0

In [15]:
cleaned_data.dropna(subset=['Description'],inplace=True)
cleaned_data['Description'].isnull().sum()

np.int64(0)

#### 将UK、U.K.统一替换成United Kingdom

In [16]:
cleaned_data['Country']=cleaned_data['Country'].replace(['UK','U.K.'],'United Kingdom')

In [17]:
print(len(cleaned_data[cleaned_data['Country']=='UK']))
print(len(cleaned_data[cleaned_data['Country']=='U.K.']))

0
0


#### 将USA替换成United States

In [18]:
cleaned_data['Country'] = cleaned_data['Country'].replace('USA','United States')

In [19]:
print(len(cleaned_data[cleaned_data['Country']=='USA']))

0


#### 将Quantity为负数的数据删除

In [20]:
cleaned_data = cleaned_data[cleaned_data['Quantity']>=0]

In [21]:
len(cleaned_data[cleaned_data['Quantity']<0])

0

#### 将UnitPrice为负数的数据删除

In [22]:
cleaned_data = cleaned_data[cleaned_data['UnitPrice']>=0]

In [23]:
len(cleaned_data[cleaned_data['UnitPrice']<0])

0

#### 将InvoiceDate的数据类型转换为日期时间

In [24]:
cleaned_data['InvoiceDate'] = pd.to_datetime(cleaned_data['InvoiceDate'])
cleaned_data['InvoiceDate']

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 530691, dtype: datetime64[us]

#### 将CustomerID的数据类型转换为字符串

In [25]:
cleaned_data['CustomerID'] = cleaned_data['CustomerID'].astype(str)
cleaned_data['CustomerID']

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 530691, dtype: str

还需要去除结尾的.0

In [26]:
cleaned_data['CustomerID'] = cleaned_data['CustomerID'].str.slice(0,-2)
cleaned_data['CustomerID']

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 530691, dtype: str

清理后的数如下：

In [27]:
cleaned_data.sample(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
14548,537603,84970L,SINGLE HEART ZINC T-LIGHT HOLDER,12,2010-12-07 12:49:00,0.95,14673,United Kingdom
430580,573585,22260,FELT EGG COSY BLUE RABBIT,4,2011-10-31 14:41:00,1.63,NaN,United Kingdom
503508,578927,85099F,JUMBO BAG STRAWBERRY,3,2011-11-27 12:26:00,2.08,15747,United Kingdom
531144,580905,23490,T-LIGHT HOLDER HANGING LOVE BIRD,6,2011-12-06 13:02:00,3.75,16313,United Kingdom
491799,578074,21889,WOODEN BOX OF DOMINOES,1,2011-11-22 16:06:00,1.25,17590,United Kingdom
497503,578400,46000S,POLYESTER FILLER PAD 40x40cm,2,2011-11-24 11:52:00,1.45,12748,United Kingdom
505443,579069,22185,SLATE TILE NATURAL HANGING,24,2011-11-28 10:31:00,1.65,14911,EIRE
131811,547649,22515,CHILDS GARDEN SPADE PINK,1,2011-03-24 12:02:00,2.10,18118,United Kingdom
363023,568531,23390,DOLLY GIRL MINI BACKPACK,2,2011-09-27 13:49:00,4.15,16713,United Kingdom
32204,539046,21212,PACK OF 72 RETROSPOT CAKE CASES,1,2010-12-15 15:50:00,0.55,17980,United Kingdom


## 保存清理后的数据

完成数据清理后，将干净整洁的数据保存到文件名为e_commerce_cleaned.csv的新文件里。

In [28]:
cleaned_data.to_csv('e_commerce_cleaned.csv', index=False)

In [29]:
pd.read_csv('e_commerce_cleaned.csv').head()

C:\Users\wesso\AppData\Local\Temp\ipykernel_29232\563775235.py:1: DtypeWarning: Columns (0: InvoiceNo) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv('e_commerce_cleaned.csv').head()


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
